substrate/frame/opf/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+					round_period_128.try_into().map_err(|_| Error::<T>::InvalidResult)?;
+				if round_infos.round_ending_block == round_infos.round_starting_block {
+					round_ending_block = round_ending_block.saturating_add(round_period.into());
+					round_infos.round_ending_block = round_ending_block;
+					round_infos.time_periods = Some(time_periods);
+				}
+			}
+			round_infos.projects_submitted =
+				projects_submitted.clone().try_into().map_err(|_| Error::<T>::InvalidResult)?;
+				VotingRounds::<T>::insert(current_round_index, round_infos); 
+
+			Self::deposit_event(Event::Projectslisted { projects_id: projects_submitted });
+			Ok(())
+		}
+
+		/// OPF Projects de-listing
+		///
+		/// ## Dispatch Origin
+		///
+		/// Must be admin origin or root
+		///
+		/// ## Details
+		///
+		/// From this extrinsic only Root can de-list a project.
+		///
+		/// ### Parameters
+		/// - `project_id`: The project that is unregistered  .
+		///
+		/// ### Errors
+		///  
+		/// ## Events
+		/// Emits [`Event::<T>::ProjectUnlisted`].
+		#[pallet::call_index(2)]
+		#[pallet::weight(<T as Config>::WeightInfo::unregister_project(T::MaxProjects::get()))]
+		pub fn unregister_project(
+			origin: OriginFor<T>,
+			project_id: ProjectId<T>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin_or_root(origin)?;
+			WhiteListedProjectAccounts::<T>::remove(&project_id);
+			Self::deposit_event(Event::<T>::ProjectUnlisted { project_id });
+
+			Ok(())
+		}
+
+		/// OPF voting logic
+		///
+		/// ## Dispatch Origin
+		///
+		/// Must be signed
+		///
+		/// ## Details
+		///
+		/// This extrinsic allows users to [vote for/nominate] a whitelisted project using their
+		/// funds. The amount defined by the user is locked and released only when the project
+		/// reward is ready for distribution, or when the project is not dimmed fundable.
+		/// Users can edit/over-write an existing vote within the vote-casting period.
+		/// At the end of the voting period, rewards are calculated based on the total user amount
+		/// attributed to each project by the user’s votes.
+		///
+		/// ### Parameters
+		/// - `project_id`: The account that will receive the reward.
+		/// - `amount`: Amount that will be locked in user’s balance to nominate a project.
+		/// - `fund`: Parameter that defines if user’s vote is in favor (*true*), or against
+		///   (*false*)
+		/// the project funding.
+		/// - `conviction`: Used to calculate the value allocated to the project, & determine
+		/// when the voter's funds will be unlocked. Amount actually locked is the amount without
+		/// conviction  
+		///
+		/// ### Errors
+		/// - [`Error::<T>::NotEnoughFunds`]: The user does not have enough balance to cast a vote
+		///  
+		/// ## Events
+		/// - [`Event::<T>::VoteCasted { who, project_id }`]: User's vote successfully submitted
+		#[pallet::call_index(3)]
+		#[pallet::weight(<T as Config>::WeightInfo::vote(T::MaxProjects::get()))]
+		pub fn vote(
```

Comment:
If we use `pallet-conviction-voting` we should use it entirely IMO, people will directly use the call in the pallet `pallet-conviction-voting`. Here we are duplicating the logic and if user calls into `pallet-conviction-voting` then stuff are inconsistent.

So this call here should not exist in this pallet. This pallet should only create the poll and get the result.

Reply:
The reason I am not directly using the conviction-pallet call is because the vote_overwriting was a functionality requested for the pallet-opf, that is not included in conviction-voting `vote` call. This is why I wrote the `try_vote` & `remove_vote` functions which includes additionnal operations to make this possible (see functions.rs), remove the old vote, and cast the new one in conviction pallet.
should we then implement vote_overwriting in conviction-voting instead?

Reply:
what do you mean by "vote_overwriting"? you mean being able to cast vote on behalf of a user? (when resubmitting vote from previous poll to next poll).

Right now if somebody dispatch a call to `pallet-conviction-voting` directly without calling from `pallet-opf` then `pallet-opf` and `pallet-conviction-voting` have inconsistent view on the votes, this doesn't look good.

Reply:
So I was asked to make it so that a voter can change his vote by voting for the same project with a different amount. It should not return and `AlreadyVoted` error, but should update all the corresponding storage.